### PR TITLE
[URGENT] Fix DDP broadcasting

### PR DESCRIPTION
**Pull request by TParcollet**
This is a serious issue. @peter842323 @Adel-Moumen  and @mravanelli - anyone technically sound should have an eye on it and review. **Our DDP, under some circumstances, is broken.** Basically, any module that does not contain trainable parameters but contains non trainable ones, will not be broadcasted properly. For BEST-RQ, for instance, it means that if you train with 4 GPUs, each GPU will get a different Quantizer. This is also true for input normalization! Up until now, if we use global as an input normalisation, the running mean and var will be calculated per GPU ... 

This PR addresses this by removing the offset to the seed of the processes. This makes sure that all processes initialise with the same seed i.e. same parameters. 

The module solution did not work as it's breaking most of the code base.

TODO:
- [x] Adel Check if it breaks stuff to redirect self.modules.module to self.modules (not good)
- [x] Adapt the input normalisation to use all_reduce operation to average the statistics.

### CODE REVIEW

File: speechbrain/core.py [hunk 1]
```diff
@@ -1715,22 +1715,20 @@ def _wrap_distributed(self):
...
+                    device_ids=None,
+                    find_unused_parameters=self.find_unused_parameters,
+                )
+            else:
+                self.modules = DDP(
+                    self.modules,
+                    device_ids=[self.device],
+                    find_unused_parameters=self.find_unused_parameters,
+                )
+            self.modules = self.modules.module
```
Comment (pplantinga) [inline speechbrain/core.py:1731]: What is the purpose of this change? from `self.modules.module` to `self.modules`

Is it related to the fix for DDP? Because I thought removing the line above:

```
                if any(p.requires_grad for p in module.parameters()):
```

was the main fix here.

Comment (TParcollet) [reply speechbrain/core.py:1731]: If you don't do this, then you can't do self.modules.Transformer. But you will need to do self.modules.module.Transformer... DDP wrapping ass an extra .module I.e refactor 100% of the recipes

Comment (TParcollet) [reply speechbrain/core.py]: We have changed of strategy (see above and commits)

**Review by Adel-Moumen — APPROVED** [review]
lgtm

File: recipes/LibriSpeech/self-supervised-learning/BEST-RQ/hparams/BEST-RQ.yaml [hunk 2]
```diff
@@ -35,7 +35,7 @@ number_of_epochs: 3000
 optimizer_step_limit: 300000
 
 # This setup is for 8 V100.
-seconds_per_batch: 400
+max_batch_len: 400 # In second and per gpu
 train_num_buckets: 150
```
Comment (pplantinga) [inline recipes/LibriSpeech/self-supervised-learning/BEST-RQ/hparams/BEST-RQ.yaml:39]: Are the BEST-RQ recipe changes related to the DDP fixes?

File: speechbrain/processing/features.py [hunk 3]
```diff
@@ -1153,6 +1154,15 @@ def forward(self, x, lengths, spk_ids=torch.tensor([]), epoch=0):
...
 
+            # If multi GPU, we need to sync the statistics across processes.
+            if is_distributed_initialized():
+                torch.distributed.all_reduce(
+                    current_mean, op=torch.distributed.ReduceOp.AVG
+                )
+                torch.distributed.all_reduce(
+                    current_std, op=torch.distributed.ReduceOp.AVG
+                )
+
```
Comment (pplantinga) [inline speechbrain/processing/features.py:1165]: I'm wondering if we should add something to the distributed tools to ensure this won't hang if its in a `if_main_process` block, similar to this broadcast code:

https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/distributed.py#L205

File: speechbrain/utils/seed.py [hunk 4]
```diff
@@ -21,7 +20,12 @@
 def seed_everything(
     seed: int = 0, verbose: bool = True, deterministic: bool = False
 ) -> int:
-    r"""Function that sets the seed for pseudo-random number generators in: torch, numpy, and Python's random module.
+    r"""Function that sets the seed for pseudo-random number generators in: torch, numpy, and Python's random module. Important note on DDP: all DDP
+    process have the same seed. This is important to ensure that parameters
+    without a require_grad set to True are the same across processes. This
+    must be taken into account if one wants to build a custom data sampler as
+    the processes would pick the same samples... SpeechBrain takes care of that
+    internally.
```
Comment (pplantinga) [inline speechbrain/utils/seed.py:28]: Have we verified that the sampler actually does work correctly when they're all the same seed? What were the original reasons for setting separate seeds? Also, do we need to provide a mechanism to access a process-specific seed?

**Review by pplantinga — COMMENTED** [review]
From the updates to the PR it looks like it is fine to only wrap modules with DDP if they have trainable parameters. I'm not sure how buffers work in pytorch, if a module consists only of e.g. batchnorm or other buffer-related blocks, would that fail here? I'm also not sure how SyncBatchNorm works, if this could somehow be leveraged for InputNorm or something like this.

Overall, I think using the same seed on the different processes is probably the way to go, I just want to confirm once more that everything is working with this

File: speechbrain/processing/features.py (hunk 3 above)
Comment (TParcollet) [reply speechbrain/processing/features.py:1165]: Oh right, good catch.

File: speechbrain/utils/seed.py (hunk 4 above)
Comment (TParcollet) [reply speechbrain/utils/seed.py:28]: The reason was that we thought it was not the case, but @Gastron  did the distributed sampler well :p

File: recipes/LibriSpeech/self-supervised-learning/BEST-RQ/hparams/BEST-RQ.yaml (hunk 2 above)
Comment (TParcollet) [reply recipes/LibriSpeech/self-supervised-learning/BEST-RQ/hparams/BEST-RQ.yaml:39]: Not really... but they are here on purpose -- these are minor "cosmetic" changes that I don't want to do a PR for..

File: speechbrain/processing/features.py (hunk 3 above)
Comment (TParcollet) [reply speechbrain/processing/features.py]: Done!

**Comment by TParcollet** [comment]
@pplantinga i've followed your advice and added a ddp_all_reduce utility in our DDP code making sure that it does not hang if called within a main_process_only context.

**Review by pplantinga — APPROVED** [review]
LGTM!
